<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span><ul class="toc-item"><li><span><a href="#Выводы" data-toc-modified-id="Выводы-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Выводы</a></span></li></ul></li><li><span><a href="#Заполнение-пропусков" data-toc-modified-id="Заполнение-пропусков-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Заполнение пропусков</a></span></li><li><span><a href="#Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Проверка-данных-на-аномалии-и-исправления.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Проверка данных на аномалии и исправления.</a></span></li><li><span><a href="#Изменение-типов-данных." data-toc-modified-id="Изменение-типов-данных.-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Изменение типов данных.</a></span></li><li><span><a href="#Удаление-дубликатов." data-toc-modified-id="Удаление-дубликатов.-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Удаление дубликатов.</a></span></li><li><span><a href="#Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Категоризация-дохода." data-toc-modified-id="Категоризация-дохода.-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Категоризация дохода.</a></span></li><li><span><a href="#Категоризация-целей-кредита." data-toc-modified-id="Категоризация-целей-кредита.-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Категоризация целей кредита.</a></span></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Ответы на вопросы.</a></span><ul class="toc-item"><li><span><a href="#Зависимость-между-количеством-детей-и-возвратом-кредита-в-срок." data-toc-modified-id="Зависимость-между-количеством-детей-и-возвратом-кредита-в-срок.-9.1"><span class="toc-item-num">9.1&nbsp;&nbsp;</span>Зависимость между количеством детей и возвратом кредита в срок.</a></span><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-9.1.1"><span class="toc-item-num">9.1.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Зависимость-между-семейным-положением-и-возвратом-кредита-в-срок." data-toc-modified-id="Зависимость-между-семейным-положением-и-возвратом-кредита-в-срок.-9.2"><span class="toc-item-num">9.2&nbsp;&nbsp;</span>Зависимость между семейным положением и возвратом кредита в срок.</a></span><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-9.2.1"><span class="toc-item-num">9.2.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок." data-toc-modified-id="Зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок.-9.3"><span class="toc-item-num">9.3&nbsp;&nbsp;</span>Зависимость между уровнем дохода и возвратом кредита в срок.</a></span><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-9.3.1"><span class="toc-item-num">9.3.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Влияние-цели-кредита-на-возврат-в-срок." data-toc-modified-id="Влияние-цели-кредита-на-возврат-в-срок.-9.4"><span class="toc-item-num">9.4&nbsp;&nbsp;</span>Влияние цели кредита на возврат в срок.</a></span><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-9.4.1"><span class="toc-item-num">9.4.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

# Исследование надёжности заёмщиков

**Цель исследования** — ответить на 4 вопроса:
1. Есть ли зависимость между количеством детей и возвратом кредита в срок?
2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
4. Как разные цели кредита влияют на его возврат в срок?

**Ход исследования**

Входные данные от банка — статистика о платёжеспособности клиентов предоставлены в файле `data.csv`. О качестве данных ничего неизвестно. Поэтому перед проверкой гипотез понадобится обзор данных.

## Обзор данных

Для начала импортируем библиотеку pandas.

In [4]:
import pandas as pd

Сохраним данные в переменную `df`.

In [5]:
df = pd.read_csv('data.csv')

Выведем на экран первые десять строк таблицы:

In [6]:
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Получим общую информацию о таблице.

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В таблице 12 столбцов с разными типами данных. 

***Согласно документации к данным:***

* `children` — количество детей в семье
* `days_employed` — общий трудовой стаж в днях
* `dob_years` — возраст клиента в годах
* `education` — уровень образования клиента
* `education_id` — идентификатор уровня образования
* `family_status` — семейное положение
* `family_status_id` — идентификатор семейного положения
* `gender` — пол клиента
* `income_type` — тип занятости
* `debt` — имел ли задолженность по возврату кредитов
* `total_income` — ежемесячный доход
* `purpose` — цель получения кредита

Названия колонок оформлены без нарушений. По столбцу `Non-Null Count` понятно, что в двух столбцах таблицы есть пропуски, значит нужно определить причину их появления, как они влияют на данные и чем их можно заполнить. Так же в глаза бросается разный стиль написания в столбце `education`. 



Посмотрим на диапазоны разброса значаний в таблице. Для этого применим к таблице метод `.desctibe`.

In [8]:
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

### Выводы

В таблице представлены данные о клиентах банка. На первый взгляд можно сказать, что данных достаточно для проверки гипотез, но в колонках `общий трудовой стаж в днях` и `ежемесячный доход` есть пропуски. Нужно определить, насколько они критичны и что с ними можно сделать.

## Заполнение пропусков

Для начала посчитаем количество пропусков в таблице:

In [9]:
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Рассмотрим строки с пропущенными значения подробнее:

In [10]:
df[df['days_employed'].isna()].head(10)

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

Пропуски в таблице относятся к количественному типу данных. Единственной закономерностью в пропущенных данных является то, что данные пропущены сразу в двух колонках одной строки. Возможно, данные в строках характеризуют клиентов без официальной работы, следовательно без официального стажа и сведений о доходе.  

Определим какую долю занимают пропуски:

In [11]:
days_employed_NaN = df['days_employed'].isna().mean()
total_income_NaN = df['total_income'].isna().mean()


9,9% информации в столбце - это пропуски. Хорошо, что для нашего исследования трудовой стаж не является ключевым показателем. Удаление строк с пропусками приведёт к потере почти 10% информации что может повлиять на полноту исследования и проверку других гипотез. 
Рассмотрим данные в столбце `days_employed` более подробно.

In [12]:
df['days_employed'].head(10)

0     -8437.673028
1     -4024.803754
2     -5623.422610
3     -4124.747207
4    340266.072047
5      -926.185831
6     -2879.202052
7      -152.779569
8     -6929.865299
9     -2188.756445
Name: days_employed, dtype: float64

Для заполнения пропусков отлично подойдёт медиана. Если взять среднее, то можно получить сильное отклонение из-за несоответствия диапазона значений и их частоты.

In [13]:
# заполним пропуски медианой.
df['days_employed'] = df['days_employed'].fillna(df['days_employed'].median())

# проверим что пропусков в столбце не осталось
print('Количество пропусков в столбце `days_employed` -', df['days_employed'].isna().sum())

Количество пропусков в столбце `days_employed` - 0


Так гораздо лучше.

Теперь разберёмся со столбцом с информацией о ежемесячном доходе: `total_income`.
Мы уже определили, что 2173 значения из 21524 это пропуски (9,9%). Пропуски так же будем заполнять медианой, т.к. это даст наиболее приближённое значение. Если выбрать среднее, то даже один клиент с очень высоким или низким доходом сместит значение. 

In [14]:
df['total_income'] = df['total_income'].fillna(df.groupby('income_type')['total_income'].transform("median"))

# проверим что пропусков в столбце не осталось
print('Количество пропусков в столбце `total_income` -', df['total_income'].isna().sum())

Количество пропусков в столбце `total_income` - 0


Пропусков больше нет, дальше проверим насколько данные близки к реальности.

## Проверка данных на аномалии и исправления.

Сразу бросается в глаза отрицательное значение стажа, исправим это.

In [15]:
# возьмём модуль числа и сохраним вместо старого хзначения.
# для того чтобы выполнение кода не остановилось встретив недопустимое значение применим внутри цикла конструкцию try...except...
for row in range(len(df)):
    try:
        df.loc[row, 'days_employed'] = abs(df.loc[row, 'days_employed'])
    except:
        print(f'Неверное значение в строке {row}')
df.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Значение в четвёртой колонке вызывает вопросы. Согласно документации, значение стажа дано в днях. Тогда клиент в четвёртой строке имеет стаж в 340266 дней! Возможно, он бессмертный, разберём подробнее.

In [16]:
# найдём количество клиентов трудовой стаж которых превышает возраст (при условии работы 365 дней в год без выходных)
df[df['days_employed'] / 365 > df['dob_years']]['days_employed'].count()

3529

Скорее всего часть данных в столбце имеет мало общего с реальностью, возможно кто-то ошибся и отделил не то количество знаков после запятой (3500 раз), возможно они были повреждены при выгрузке. В любом случае неверных значений оказалось, больше чем пропусков. Запросить свежую выгрузку у банка не представляется возможным, т.к. это учебный проект.

Посмотрим какие уникальные значения есть в столбце с детьми:

In [17]:
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

В результате мы видим, что в количестве детей в семье однозначно есть прогрессия. Клиентов с большим количеством детей меньше. Но есть и аномалии, это отрицательные и очень высокие значения. К счастью их, немного и они не вносят большой погрешности. Так как 20 детей, это уникальные случаи, то будем считать что это опечатка от 2, как и в случае с -1 к 1.

In [18]:
df.loc[df['children'] == -1, 'children'] = 1
df.loc[df['children'] == 20, 'children'] = 2
df['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

Проверим на наличие уникальных значение столбец `gender`

In [19]:
df['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

Female, Male и скорее всего Хитмен. Посчитаем количество значений `XNA`.

In [20]:
df[df['gender'] == 'XNA']['gender'].count()

1

Он такой один. Или она. Определить пол по имеющимся данным не получится, поэтому одной строчкой придётся пожертвовать.

In [21]:
display(df[df['gender'] == 'XNA'])

children  days_employed  dob_years            education  education_id  \
10701         0    2358.600502         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10701  гражданский брак                 1    XNA   компаньон     0   

        total_income               purpose  
10701  203905.157261  покупка недвижимости

In [22]:
df.drop(labels = [10701],axis = 0, inplace = True)

In [23]:
df['gender'].unique()

array(['F', 'M'], dtype=object)

## Изменение типов данных.

Так же значения в колонке имеют десятичную часть до шестого знака, нам такая точность излишне, преобразуем в целое число.

In [24]:
#Приведём данные к целочисленному типу
df['days_employed'] = df['days_employed'].astype('int')
df['total_income'] = df['total_income'].astype('int')

## Удаление дубликатов.

В столбце `education` встречается написание одного значения разным стилем. Получим список всех уникальных значений столбца.

In [25]:
df['education'].unique()

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

Получается, что каждое образование записано тремя альтернативными способами. Скорее всего это связано с тем, что нет общепринятого стандарта как заносить данные в таблицу и люди записывают их так как им удобно. Группировка или сортировка такого столбца будет некорректной, приведём всё в порядок. Для этого приведём все символы строк к нижнему регистру.

In [26]:
df['education'] = df['education'].str.lower()
df['education'].unique() # ещё раз запросим список уникальных значений

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

Проверим датафрейм на наличие явных дубликатов:

In [27]:
df_duplicated = df[df.duplicated()]
print('Явных дубликатов -', df.duplicated().sum())
df_duplicated.head(10)

Явных дубликатов - 71


children  days_employed  dob_years education  education_id  \
2849         0           1203         41   среднее             1   
3290         0           1203         58   среднее             1   
4182         1           1203         34    высшее             0   
4851         0           1203         60   среднее             1   
5557         0           1203         58   среднее             1   
6312         0           1203         30   среднее             1   
7808         0           1203         57   среднее             1   
7921         0           1203         64    высшее             0   
7938         0           1203         71   среднее             1   
8583         0           1203         58    высшее             0   

              family_status  family_status_id gender income_type  debt  \
2849        женат / замужем                 0      F   сотрудник     0   
3290       гражданский брак                 1      F   пенсионер     0   
4182       гражданский брак                 1      F   сотрудник     0   
4851       гражданский брак                 1      F   пенсионер     0   
5557       гражданский брак                 1      F   пенсионер     0   
6312        женат / замужем                 0      M   сотрудник     0   
7808       гражданский брак                 1      F   пенсионер     0   
7921       гражданский брак                 1      F   пенсионер     0   
7938       гражданский брак                 1      F   пенсионер     0   
8583  Не женат / не замужем                 4      F   пенсионер     0   

      total_income                           purpose  
2849        142594           покупка жилья для семьи  
3290        118514                   сыграть свадьбу  
4182        142594                           свадьба  
4851        118514                           свадьба  
5557        118514                   сыграть свадьбу  
6312        142594  строительство жилой недвижимости  
7808        118514             на проведение свадьбы  
7921        118514             на проведение свадьбы  
7938        118514             на проведение свадьбы  
8583        118514        дополнительное образование

Метод `.duplicated()` выявил 405 дубликат в датафрейме.
Просмотрев таблицу с дубликатами, можно сказать, что дубликатов в ней нет. Есть похожие клиенты, но с разным возростом, т.е. люди могли обращаться в банк в разные периоды жизни или с разными целями.

После предобработки датафрейм не содержит пропусков, дубликатов и аномалий. Теперь можно приступать к дальнейшему формированию и непосредственно аналитике.

## Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

Каждому значению в `education` соответствует значение в столбце `education_id`, аналогично для столбца `family_status`. Для удобства работы с таблицей и её большей наглядности создадим из этих столбцов отдельные словари.

Для начала найдём все уникальные значения столбца `education`:

In [28]:
df['education'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      743
начальное                282
ученая степень             6
Name: education, dtype: int64

Выделим столбцы в отдельный словарь:

In [29]:
education_dict = df[['education_id','education']]
education_dict.head()

education_id education
0             0    высшее
1             1   среднее
2             1   среднее
3             1   среднее
4             1   среднее

В словаре большое количество дубликатов. Фактически их можно сократить до 5. Удалим дубликаты и сбросим индексы.

In [30]:
education_dict = education_dict.drop_duplicates().reset_index(drop=True)
education_dict

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

Теперь столбец `education` можно удалить и обращаться к его данным по индексу из `education_id`. Поступим аналогично с `family_status`:

In [31]:
family_status_dict = df[['family_status_id', 'family_status']]
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True)
family_status_dict.head()

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  Не женат / не замужем

Столбцы `education` и `family_status` можно удалить из исходного датафрейма.

In [32]:
df = df.drop(columns=['education', 'family_status'], axis = 1)
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0         340266         53             1                 1      F   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу

## Категоризация дохода.

Посчитаем клиентов по уровню зарплаты:

In [33]:
df['total_income'].value_counts()

142594    1105
172357     509
118514     414
150447     147
161807       3
          ... 
300306       1
381117       1
70100        1
110461       1
82047        1
Name: total_income, Length: 18607, dtype: int64

Сложно сделать какие-либо выводы из единичных отрывков данных. Для того что бы дать аналитику по данному столбцу, сгруппируем данные по диапазонам зарплат.

In [34]:
# напишем функцию возвращающую категорию зависящую от диапазона
def income_group(income):
    if 0 <= income <= 30000:
        return 'E'
    if 30001 <= income <= 50000:
        return 'D'
    if 50001 <= income <= 200000:
        return 'C'
    if 200001 <= income <= 1000000:
        return 'B'
    if 1000001 <= income:
        return 'A'
    
#проверим
print(income_group(8000000))
print(income_group(235000))
print(income_group(60000))
print(income_group(50000))
print(income_group(25000))

A
B
C
D
E


Функция работает корректно. Теперь методом `.apply` добавим новый столбец с полученными категориями в таблицу:

In [35]:
df['total_income_category'] = df['total_income'].apply(income_group)
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0         340266         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C

Результатом работы стал столбец с категориями клиентов. Категорий всего 5, в таком виде данные подходят для анализа.

## Категоризация целей кредита.

Создайте функцию, которая на основании данных из столбца purpose сформирует новый столбец purpose_category, в который войдут следующие категории:
'операции с автомобилем',
'операции с недвижимостью',
'проведение свадьбы',
'получение образования'.
Например, если в столбце purpose находится подстрока 'на покупку автомобиля', то в столбце purpose_category должна появиться строка 'операции с автомобилем'.
Вы можете использовать собственную функцию и метод apply(). Изучите данные в столбце purpose и определите, какие подстроки помогут вам правильно определить категорию.

In [36]:
df['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

In [37]:
def purpose_cat(purpose):
    if 'жил' in purpose or 'движ' in purpose:
        return 'операции с недвижимостью'
    elif 'авто' in purpose:
        return 'операции с автомобилем'
    elif 'образован' in purpose:
        return 'получение образования'
    elif 'свадьб' in purpose:
        return 'проведение свадьбы'
    else: 
        return 'другие цели'
print(purpose_cat('покупка жилья'))
print(purpose_cat('приобретение автомобиля'))
print(purpose_cat('дополнительное образование'))
print(purpose_cat('сыграть свадьбу'))
print(purpose_cat('купить бананы'))

операции с недвижимостью
операции с автомобилем
получение образования
проведение свадьбы
другие цели


In [38]:
df['purpose_category'] = df['purpose'].apply(purpose_cat)
df['purpose_category'].unique()

array(['операции с недвижимостью', 'операции с автомобилем',
       'получение образования', 'проведение свадьбы'], dtype=object)

## Ответы на вопросы.

В результате получаем таблицу готовую к анализу и можем приступить к ответам на вопросы.

Для начала выведем итоговую таблицу чтобы она была перед глазами и было удобнее работать с данными.

In [39]:
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0         340266         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы

### Зависимость между количеством детей и возвратом кредита в срок.


Для начала сформируем сводную таблицу и добавим к ней столбец с суммой значений строк.

In [40]:
df_pivot_children = df.pivot_table(index=['children'], columns='purpose_category', values='debt', aggfunc='mean')
df_pivot_children['Общий итог'] = df_pivot_children.sum(axis=1)
df_pivot_children

purpose_category  операции с автомобилем  операции с недвижимостью  \
children                                                             
0                               0.085203                  0.067080   
1                               0.106557                  0.080775   
2                               0.120482                  0.083955   
3                               0.083333                  0.076923   
4                               0.100000                  0.142857   
5                               0.000000                  0.000000   

purpose_category  получение образования  проведение свадьбы  Общий итог  
children                                                                 
0                              0.086448            0.074146    0.312877  
1                              0.102623            0.095149    0.385105  
2                              0.112172            0.066667    0.383276  
3                              0.057971            0.156250    0.374477  
4                              0.000000            0.000000    0.242857  
5                              0.000000            0.000000    0.000000

В результате мы видим, что количество невозвратов кредита в срок от количества детей немного разнятся. Сгрупперуем данные по формуле: количество клиентов без детей и с долгом / (количество клиентов без детей и с долгом + количество клиентов без детей и без долга) и упорядочим по убыванию коэффициента.

In [41]:
df_pivot_children = df.groupby(['children']).agg({'debt':'sum'})
df_pivot_children['Надёжность от количества детей'] = round(
    df_pivot_children['debt'] /
    (df_pivot_children['debt'] + 
     df['children'].value_counts() - df_pivot_children['debt'])
, 3)
df_pivot_children = df_pivot_children.sort_values(by='Надёжность от количества детей', ascending=False)

df_pivot_children

debt  Надёжность от количества детей
children                                      
4            4                           0.098
2          202                           0.095
1          445                           0.091
3           27                           0.082
0         1063                           0.075
5            0                           0.000

#### Вывод:

Из получившейся таблицы мы можем сделать вывод что есть зависимость вероятность просрочить кредит от числа детей. На первом месте по просроченным кредитам клиенты с 4 детьми. На последнем клиенты без детей, это также самая объёмная категория клиентов. Надёжнее всех оказались клиенты с 5 детьми т.к. невозвраты кредита в срок в этой категории отсутсвуют.

### Зависимость между семейным положением и возвратом кредита в срок.

In [42]:
df_pivot_family_status = df.groupby(['family_status_id']).agg({'debt':'sum'})
df_pivot_family_status['Надёжность от семейного положения'] = round(
    df_pivot_family_status['debt'] /
    (df_pivot_family_status['debt'] + 
     df['family_status_id'].value_counts() - df_pivot_family_status['debt'])
, 3)
df_pivot_family_status = df_pivot_family_status.merge(family_status_dict, on='family_status_id')
df_pivot_family_status = df_pivot_family_status.sort_values(by='Надёжность от семейного положения', ascending=False)

df_pivot_family_status

family_status_id  debt  Надёжность от семейного положения  \
4                 4   274                              0.097   
1                 1   388                              0.093   
0                 0   931                              0.075   
3                 3    85                              0.071   
2                 2    63                              0.066   

           family_status  
4  Не женат / не замужем  
1       гражданский брак  
0        женат / замужем  
3              в разводе  
2         вдовец / вдова

#### Вывод:

В результате мы видим что клиенты в категории `вдовец / вдова` имеют самый низкий процент невозврата кредита в срок. Чуть меньше процент в категории `в разводе`.
За ними идут клиенты в категории `женат / замужем`.
Категории `гражданский брак` и `Не женат / не замужем` имеют самый высокую вероятность невозврата в срок. 


### Зависимость между уровнем дохода и возвратом кредита в срок.

In [43]:
df_pivot_income = df.groupby(['total_income_category']).agg({'debt':'sum'})
df_pivot_income['Надёжность от уровня дохода'] = round(
    df_pivot_income['debt'] /
    (df_pivot_income['debt'] + 
     df['total_income_category'].value_counts() - df_pivot_income['debt'])
, 3)
df_pivot_income = df_pivot_income.sort_values(by='Надёжность от уровня дохода', ascending=False)

df_pivot_income

debt  Надёжность от уровня дохода
total_income_category                                   
E                         2                        0.091
C                      1360                        0.085
A                         2                        0.080
B                       356                        0.071
D                        21                        0.060

#### Вывод:

Надёжнее всех оказались клиенты в категории в категории `D` - это клиенты с доходом в диапазоне 30001–50000.
Самый высокий процент невозвратов у категории `E` - категории с самым низким доходом.

### Влияние цели кредита на возврат в срок.

In [44]:
df_pivot_purpose = df.groupby(['purpose_category']).agg({'debt':'sum'})
df_pivot_purpose['Надёжность от цели кредита'] = round(
    df_pivot_purpose['debt'] /
    (df_pivot_purpose['debt'] + 
     df['purpose_category'].value_counts() - df_pivot_purpose['debt'])
, 3)
df_pivot_purpose = df_pivot_purpose.sort_values(by='Надёжность от цели кредита', ascending=False)

df_pivot_purpose

debt  Надёжность от цели кредита
purpose_category                                          
операции с автомобилем     403                       0.093
получение образования      370                       0.092
проведение свадьбы         186                       0.079
операции с недвижимостью   782                       0.072

#### Вывод:

Из полученных данных видно что в основном клиенты берут заёмные средства для операций с недвижимостью. В этой же категории самый низкий процент невозвратов. За ними идут те, кто взял кредит на проведение свадьбы. Самыми не надёжными оказались те, кто берёт заёмные средства для получения образования и операций с авто.

## Общий вывод:

В результате расчётом мы получили следующие характеристики клиентов:

In [45]:
columns = ['Количество детей', 'Семейное положение', 'Уровень дохода', 'Цель кредита']

data = [['4', 'Не женат / не замужем', 'E', 'операции с автомобилем'],
        ['1', 'гражданский брак','C','получение образования'],
        ['0', 'женат / замужем', 'A', 'проведение свадьбы'],
        ['3', 'в разводе', 'B', 'операции с недвижимостью'],
        ['2', 'вдовец / вдова', 'D', '-']]

final_data = pd.DataFrame(data=data, columns=columns)
final_data

Количество детей     Семейное положение Уровень дохода  \
0                4  Не женат / не замужем              E   
1                1       гражданский брак              C   
2                0        женат / замужем              A   
3                3              в разводе              B   
4                2         вдовец / вдова              D   

               Цель кредита  
0    операции с автомобилем  
1     получение образования  
2        проведение свадьбы  
3  операции с недвижимостью  
4                         -

Так самым ненадёжным оказывается многодетный клиент с низким уровнем дохода, не состоящий в браке, берущий деньги для операций с автомобилем.